core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -545,61 +546,130 @@ object Validator {
+          "An Interface type must define one or more fields"
         )
+      case Some(fields) => validateFields(fields)
+    }
+  }
+
+  private def onlyInputType(`type`: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsInputType()
+    def isInputType(t: __Type): Either[__Type, Unit] = {
+      import __TypeKind._
+      t.kind match {
+        case LIST | NON_NULL              => t.ofType.fold[Either[__Type, Unit]](Left(t))(isInputType)
+        case SCALAR | ENUM | INPUT_OBJECT => Right(())
+        case _                            => Left(t)
+      }
+    }
 
-    def doesNotStartWithUnderscore(field: __InputValue): IO[ValidationError, Unit] =
-      IO.when(field.name.startsWith("__"))(
+    IO.whenCase(isInputType(`type`)) {
+      case Left(errorType) =>
         failValidation(
-          s"InputObject can't start with '__': ${field.name}",
-          """The input field must not have a name which begins with the
-characters {"__"} (two underscores)"""
+          s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an InputType",
+          """The input field must accept a type where IsInputType(type) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
         )
-      )
+    }
+  }
 
-    def onlyInputFieldType(field: __InputValue): IO[ValidationError, Unit] =
-      IO.whenCase(isInputType(field.`type`())) {
-        case Left(errorType) =>
-          failValidation(
-            s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an InputType",
-            """The input field must accept a type where IsInputType(inputFieldType) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
-          )
+  private def onlyOutputType(`type`: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsOutputType()
+    def isOutputType(t: __Type): Either[__Type, Unit] = {
+      import __TypeKind._
+      t.kind match {
+        case LIST | NON_NULL                            => t.ofType.fold[Either[__Type, Unit]](Left(t))(isOutputType)
+        case SCALAR | OBJECT | INTERFACE | UNION | ENUM => Right(())
+        case _                                          => Left(t)
       }
+    }
 
-    t.inputFields match {
-      case None | Some(Nil) =>
+    IO.whenCase(isOutputType(`type`)) {
+      case Left(errorType) =>
         failValidation(
-          s"InputObject ${t.name.getOrElse("")} does not have fields",
-          "An Input Object type must define one or more input fields"
+          s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an OutputType",
+          """The input field must accept a type where IsOutputType(type) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
         )
-      case Some(fields) => validateFields(fields)
     }
   }
 
+  private def noDuplicateFieldName(fields: List[__Field]) = {
```

Comment:
Maybe should be called `noDuplicateFieldInterface` since the same will be needed for object types. 
Or move it inside the function the validates interfaces to avoid confusion.

Reply:
moved inside the interface validation method